### DISCUSSION TITLE
GLM and API surface area

### DISCUSSION THREAD

**Discussion by wtholliday** [elalish/manifold#962] [General]
I've put in a few hours of work trying to replace GLM. It's not exactly a couple hundred lines as @pca006132 suggested earlier. And manifold doesn't use GLM in totally trivial ways. There is more numerical code in Manifold that I expected (`svd.h`, `smoothing.cpp`). There's a bit of stuff for me to get wrong. To be clear, it's totally something I can do and have done this sort of thing before (though I haven't had to write my own vector/matrix classes since grad school).

So I'm wondering, should GLM just be out of the public API but still used for internal calculations? Since it's header-only what's the harm in just checking it in to the repo so it doesn't need to be an external dep? That seems preferable to some little short-on-features set of structs I would write. Why reinvent the wheel?

**Comment by elalish** [comment]
Great question. I really don't know - I originally used GLM as a submodule for pretty much exactly this reason. I'm still trying to understand how dependency management in C++ can be so bad that I can't even use a header math library! @fire, @kintel, @starseeker, @pca006132 let's see if we can put our heads together and figure out a decent solution here.

Things we have tried:
1) GLM as a submodule (apparently everyone hates submodules)
2) link in libglm or else auto-download it (current state)
3) allow auto-download to be disabled in CMake (#961)

Possible futures:

3) Dump giant pile of GLM headers into our own repo, effectively forking it?
4) Trim down giant header pile into what we actually use - fork it even more.
5) Make our own math library (maybe based on GLM) - completely fork it.
6) Leave GLM as is but remove it from our public API - a bit like our C bindings.
7) ... any other ideas?

Please chime in and help us come to consensus here.

**Comment by pca006132** [comment]
Ignoring APIs that are provided by `std`, some constants as well as matrix/vector multiplication and quaternion operations, here is a list of APIs that we use:

```
glm::all
glm::any
glm::clamp
glm::conjugate
glm::cross
glm::dot
glm::greaterThan
glm::greaterThanEqual
glm::length
glm::lessThan
glm::max
glm::min
glm::mix
glm::mod
glm::normalize
glm::outerProduct
glm::quat_cast
glm::rotate
glm::rotateX
glm::rotateY
glm::rotateZ
glm::rotation
glm::saturate
glm::scale
glm::translate
glm::transpose

glm::determinant
glm::inverse
glm::smoothstep
glm::lerp
```

I would expect the first part to be easy to implement and the remaining 5 requiring more code. But yeah, I was probably underestimating the code needed...

Regarding should we keep glm: I don't think removing it from the public API will buy us anything for godot, they care about compile-time dependency as well. I think it is fine to keep this as is, if someone wants to get rid of this they are free to open a PR and implement everything...

**Comment by starseeker** [comment]
My take would be to avoid re-inventing the glm wheel and stick with solution 2/3.  That said, I would *definitely* advise getting glm out of the public manifold API (option 6), unless there are serious negative implications in performance or code complexity in doing so.

I'll own up to a selfish reason for wanting option 6 - BRL-CAD bundles manifold, and its exposure of glm in the public headers means we also need to bundle glm because of how we manage and build against dependencies.  It would be better if glm were just a compile time dependency - we do distinguish between them.

More importantly, if someday down the road glm becomes problematic for whatever reason and Manifold wants or needs to switch to some other spiffy math library, having glm types in the public headers means that'll be a breaking change for client codes on a much larger scale.

**Reply by elalish** [reply]
Now that your big restructure is merged, would you be interested in trying a PR for option 6? If we're going to it, it should be now for v3.0.

**Reply by starseeker** [reply]
Um.  I'm game, but I might need some assistance with the right way to efficiently translate data and hide types for something like that.  I'll take a look tomorrow and see what I can figure out.

**Reply by starseeker** [reply]
Ooof.  OK, so as a first cut at this I'm looking at what's involved with pushing glm down below the public headers.  If I make (most of) common.h private, make the vec2, etc. uses into their own types and make glm:: type usage explicit, and start propagating that through the code I'm seeing some progress, but the Box and Rect types look to be problem children - they've got a lot of methods that are returning vector types.

One option would be to have all methods return the public manifold types, and just translate "under the hood" in an implementation .cpp file where the math happens, but if the client codes using Manifold don't need most of those operations - i.e. they're "implementation use only" even if they are technically public API - then we'll be translating the types without any real benefit.

The next option that springs to mind is to have a "pointer to implementation" approach where any guts that we don't really want to expose are hidden, a.l.a https://github.com/BRL-CAD/brlcad/tree/main/doc/notes/c_cxx_patterns/05_PImpl_CXX (although in this case it'd be all C++).  If we *do* want all the Box and Rect methods visible for calling codes it's moot, but if we want to hide implementation details of otherwise public classes that's a possible approach.

Honestly my instinct is to see how much of this can be hidden while still presenting the public containers and methods needed, but that might wind up conflicting with design goals (particularly for the web-based Python/wasm environments) so I wanted to see what your thoughts are.

**Reply by elalish** [reply]
So, my original reasoning for exposing GLM vectors (naively, many years ago) was "it's so convenient to have all these operations already defined for these vectors, why would anyone *not* want GLM in their code?". If I'm understanding the feedback from @fire, there are enough different little math libraries out there that everyone would rather just use their own instead of being stuck with our choice. Of course that implies doing two conversions: from GLM to generic struct in our code and from generic struct to their own math library in theirs. I'm not too worried about the cost - these are tiny structs, and the big vectors of points have already been converted to flat generic lists. 

So, I think it's in keeping with the transition from `Mesh` to `MeshGL` to switch from exposing e.g. `glm::vec3` to converting it to something generic - maybe just `std::array<float, 3>`? I usually gauge whether my API change is good by whether it simplifies our tests - this will not. But I was used to assuming the use of GLM in our tests, which maybe isn't fair. What say our other users? Pinging @stephomi, @geoffder, @stewartoallen.

**Reply by elalish** [reply]
One other naive C++ question: is there a vec3 struct that we could safely cast a `glm::vec3` struct back and forth to? So, don't actually touch the memory, but just cast away the GLM-ness of it? Then I could cast them back to GLM in our tests and it would all be pretty transparent. Not sure if that's a "there be dragons" scenario though.

**Reply by starseeker** [reply]
Four conversions, actually - from client code's type to the manifold public type, manifold public type to internal processing type (currently glm), from glm back to manifold public type, and from manifold public type to client code type.

Copying is typically an issue when the meshes being processed are so large (or you're working with so many at once) that you start to push the limits of memory.  That's why some libraries go whole hog with templates  and adapters (https://github.com/jlblancoc/nanoflann does that, IIRC) but I don't know if that's practical for Manifold or not.

**Reply by stephomi** [reply]
I can only speak for the C++ interface but yes, getting rid of dependencies from the public headers is a good thing.

> big vectors of points have already been converted to flat generic lists.

That's the most important thing.
I could nitpick and says that Manifold still requires a copy since it relies on std::vector but it's not really relevant imo since the boolean operation is going to be the bottleneck anyway.

> One other naive C++ question: is there a vec3 struct that we could safely cast a glm::vec3 struct back and forth to? So, don't actually touch the memory, but just cast away the GLM-ness of it

Is the copy really an issue?
Probably best to assume the manifold user will not have glm and will have to copy anyway.

---
I'm not sure it's a good idea but if you really want to keep a glm API, you can also omit the glm include in `manifold.h` and declare the functions as template
`template<typename V> void Translate(V);`
And in one of the manifold cpp, instanciate it with:
`template void Translate(glm::vec3);`

**Reply by kintel** [reply]
> One other naive C++ question: is there a vec3 struct that we could safely cast a `glm::vec3` struct back and forth to? So, don't actually touch the memory, but just cast away the GLM-ness of it? Then I could cast them back to GLM in our tests and it would all be pretty transparent. Not sure if that's a "there be dragons" scenario though.

From past experience, one challenge could be memory alignment requirements if you wish to do SIMD processing of elements. Looking at Eigen, which implements SIMD, could reveal some solutions to that.

**Reply by starseeker** [reply]
In case it's helpful, one of things I've tried was expanding out the use statements in common.h that abbreviated the glm types, to make it clearer where that API is in use:  https://github.com/starseeker/manifold/tree/glmexplicit

**Reply by pca006132** [reply]
If `Box` and `Rect` are the only problems, we can just have a custom implementation of addition/max with `std::array<double, N>`? It just takes a few lines.

I don't think having a pointer to implementation is good for these common types, iirc we use them quite often in `Collider`, which the performance is quite important.

**Reply by starseeker** [reply]
Possibly - I'll have to admit my primary coding style is C-ish with a smattering of C++ where convenient, so I'm not the best person to suggest possibilities for how C++ types could be used here, and I don't have a good feel for how various class design structures would impact performance.

My naive notion was that if we just have some of the class contents and methods effectively be "public only within Manifold itself" and not visible to calling applications we might be able to hid glm as an implementation detail more easily.  That's why I was asking what functionality needed to be exposed via the various APIs - if some of the user facing features need more complex types, then that ups the requirements.  For example, if Box and Rect are used internally but we can just return the arrays of values in the public Manifold methods, then Box and Rect can stay as-is and be private to Manifold...

**Reply by pca006132** [reply]
I'm fine to making `Box` and `Rect` private within manifold. In fact, I think we probably should clean up `common.h` to only keep the essential things, e.g. not sure if `CCW` should stay there, as well as some debug helpers.

**Reply by stephomi** [reply]
Yes it’s mostly about getting rid of the glm include in the headers exposed to the user (at least the mandatory ones, manifold.h).
Internal use is fine.

It did bite me a while ago, back then I was using the “master” version of glm but manifold was using the “release” glm version, which was much more older.

Box is only used once in manifold.h, so you can simply return an array and keep the glm box in internal headers.

**Reply by starseeker** [reply]
@pca006132 I like the idea of cleaning up common.h (and the other public headers) so they only expose what we want to expose as the public API - that'll help a lot.

That's probably the first step then, rather than me trying to redefine vec2 and friends straight away - no point in recoding things to use different vec/mat types if they wind up being internal.  How should that work?

**Reply by pca006132** [reply]
Sounds good, I can open a PR for some simple cleanup first. I think changing vec/mat to array types will be a rather large change, because we probably need to change some of our tests and bindings as well, so that should go into another PR.

**Reply by starseeker** [reply]
This might be a crazy question, but does the Manifold class need to be public?  If MeshGL is how data goes into and out of Manifold, could that be the only public class?  The comment in manifold.h says Manifold is the "library's internal representation of an oriented 2-manifold, triangle mesh"...

**Reply by pca006132** [reply]
The actual internal representation is a lot more complicated than that, I think the documentation was outdated for 2 years or something.

It should be public because something the user can operate on, and easily copy without any cost (it is just a shared ptr). And users may want to query it for something without actually exporting the mesh, e.g. if they want some approximation of the bounding box. (not implemented yet)

**Comment by kintel** [comment]
..or go straight for Eigen, which may be a bit more ubiquitous :)

**Reply by elalish** [reply]
Eigen feels like massive overkill for what we need - it's got to be 100x the size of GLM, right?

**Reply by kintel** [reply]
Yeah, probably, but mostly (only?) header only if that helps.

**Reply by elalish** [reply]
Well, GLM is already header-only. If I start needing higher dimensions than 4 I'll grab Eigen, but I don't think that'll happen in this library.

**Comment by wtholliday** [comment]
In the interest of incremental progress, I'd say doing option 6 is a good one. That reduces the API surface area. The decision of whether to (effectively) fork GLM is a separate issue.

As an alternative to GLM, maybe this one? https://github.com/sgorsten/linalg. Just one header so you can drop it into the repo without much guilt :)

**Reply by elalish** [reply]
linalg looks pretty slick, actually! If you'd like to take a look at the feasibility of switching to that from GLM, I'd be quite interested. 

As you say, I suppose option 6 is still a good idea regardless.

**Reply by starseeker** [reply]
Maybe I'm unnecessarily paranoid, but when we talk about switching out glm I can't help thinking back to https://github.com/elalish/manifold/pull/610 and how that one line change made such a difference in the reliability of the booleans (and I think there have been a few more like that for zebra et. al.)  I guess the test suite will have our backs if we did make such a change...

**Reply by elalish** [reply]
Appreciated, but that was my own code deep in our triangulator, which is always our problem child. These vector ops are much more settled territory, and every one of those math functions gets called probably no less than millions of times in our test suite. I think this should be fairly safe - more a matter of how many functions aren't included in the other library. And GLM has an open enough license we can grab a few lines of code if we need them.

**Reply by fire** [reply]
https://github.com/sgorsten/linalg looks slick, bit worried about the amount of work we have to do but `Make our own math library (maybe based on GLM) - completely fork it.` and using linalg seem to be in the same area.

**Comment by elalish** [comment]
Starting a new thread based on:

>One other naive C++ question: is there a vec3 struct that we could safely cast a glm::vec3 struct back and forth to? So, don't actually touch the memory, but just cast away the GLM-ness of it? Then I could cast them back to GLM in our tests and it would all be pretty transparent. Not sure if that's a "there be dragons" scenario though.

According to [this](https://stackoverflow.com/questions/21858311/c-cast-float-to-glmvec3), GLM is actually designed for this kind of casting. Would this be a simple solution to 6)? Just cast any GLM structs currently exposed in our public API to generic arrays? What do you think @pca006132?

**Reply by pca006132** [reply]
I think the problem is for things like polygon, where we have a vector of vec2, where the thing is nested. Need to check if glm guarantees alignment is the same as normal double, and if std::array has special alignment requirements, as well as whether this kind of type cast is ub (c++ is really strict about this even for things that are compatible in terms of memory layout)

**Reply by wtholliday** [reply]
I don't think casting and data layout should be an issue. Do the conversions as necessary when you red/write from arrays in the API, much like how you'd convert from packed types to simd types in a shader.

**Reply by elalish** [reply]
If it makes things easier I'm also open to changing `Polygon` in a similar way as `Mesh` became `MeshGL` - replace the arrays of GLM structs with flat arrays. That would be more consistent anyway.

**Reply by pca006132** [reply]
> I don't think casting and data layout should be an issue. Do the conversions as necessary when you red/write from arrays in the API, much like how you'd convert from packed types to simd types in a shader.

This is fine, but this essentially pushes the changes deep inside our code, which I wanted to avoid.

> If it makes things easier I'm also open to changing `Polygon` in a similar way as `Mesh` became `MeshGL` - replace the arrays of GLM structs with flat arrays. That would be more consistent anyway.

Sounds good. I think this and `Hull` are the only places where we have a `std::vector` of glm types? If we change all these into flat arrays, we don't have to think about casting. For the other APIs, conversion between glm types and arrays are simple and cheap (either casting or copying 3 elements).

**Reply by elalish** [reply]
That sounds perfect!

**Reply by starseeker** [reply]
Being selfish again, my input data to Manifold is a C array of doubles holding vertex coordinates and a C array of triangle vertex indices.  At the moment I'm using the older Mesh container and creating glm::vec3 and glm::ivec3 copies of the data to populate it, so I know I'm going to need to update things in any case.  Since I'll be updating that code, I can't help wondering...  In an ideal world it would be awesome to create C++ containers (std::array?) from those C arrays and pass them off directly to Manifold with no further processing - would that be possible?

**Reply by pca006132** [reply]
`MeshGL64` already accepts flattened array of doubles and triangle vertex indices. It may be possible to make our API accept a `VecView` and avoid the copy to a C++ vector, but this may require duplicating some code on our side. Can have a look...

**Reply by wtholliday** [reply]
@starseeker what's your concern there? extra memory usage? eliminating the conversion code? (isn't it just a few lines?)

**Reply by starseeker** [reply]
@wtholliday Primarily extra memory usage.

**Comment by pca006132** [comment]
Opening a new thread regarding avoiding copy for `MeshGL/MeshGL64`. I think this is a fair request because we usually don't need to use the arrays inside as vectors, it is just some temporary data structure, so allowing users to provide pointers to some C array can make sense.

My proposal for this would be to make `MeshGLP` self-referential, i.e. things like `vertProperties` can point to a vector stored in `dataSource`, or just a pointer to some user provided buffer.:

```c++
struct MeshGLP {
  VecView<Precision> vertProperties;
  // ...
  std::unique_ptr<DataSource<Precision, I>> dataSource = nullptr;
};
```

The data source pointer handles ownership. When `MeshGLP` is destroyed, we will free the data source if it is not null, and it will free the buffers inside if any. Typically, for `MeshGLP` exported from manifold, users don't have to worry about lifetime because it is the same as the existing API, users can deallocate the `Manifold` object without worrying about dangling pointers in this `MeshGLP` object. However, if users set the fields of `MeshGLP` for import into manifold and don't explicitly copy things into the data source, they have to take care of the lifetime themselves, and not delete the referred vector/array before constructing a manifold. After a manifold is being constructed, they don't have to worry about the lifetime of data inside the used `MeshGLP` as well.

**Reply by pca006132** [reply]
Hmm, probably our test code is the most affected if we want to do this, because we directly assign a vector to it and expect it to take ownership. But the changes are quite trivial. Not sure if this will increase the difficulty of using the library...

**Reply by elalish** [reply]
In general I'm not super persuaded by the folks who are desperate to avoid an extra copy of their verts/tris. That is such a tiny fraction of the time/memory used just in our `Manifold` constructor, much less the Booleans, that it seems silly to worry about. I'm more interested in library usability and avoiding foot-guns. So, if this change makes our tests simpler / easier to understand, then I'll be in favor. @starseeker you're welcome to convince me otherwise.

**Reply by pca006132** [reply]
Yeah, I think this will not help to make things more readabld. I am interested in this mainly due to PL issue, this cannot be expressed with lifetime in rust :P

**Reply by starseeker** [reply]
@elalish For the use cases I typically encounter day to day, I'm also not really desperate to avoid the extra copy - if the decision is made that it doesn't make sense, that's OK.  Where it can potentially matter is if you end up with input meshes so large they start pushing your available system RAM.  As a contrived academic example, how many rotated lucy models could we union together with Manifold?

However, if internally Manifold's storage requirements are large compared to the initial size of the import mesh, then I agree that data size savings probably wouldn't be sufficient to warrant increased API complexity/ugliness just for that purpose.

**Reply by elalish** [reply]
Yeah, if we want to optimize memory usage, there's lots more fertile ground in our `Impl` than our public API.

**Reply by starseeker** [reply]
The advantage of setting up the public API now to support it (if that can be done with minimal cost) is we then wouldn't have to change the public API down the road if memory optimization did get more aggressive...

**Reply by elalish** [reply]
Pinging @ncthbrt @NickUfer since they are interested in Rust bindings and this potential API change might make a difference there. I'm still a little hazy on best practices here for passing large arrays in and out of a library. What do other popular libraries do? In general I feel like a functional approach is nice where ownership isn't ever being passed around.

**Reply by hrgdavor** [reply]
I was guilty many times looking into optimisation too soon. So at first this sounded like nice to have.

Thinking about it more I feel I would want to have safer version too if optimised version is present. This then pushes me into direction of preferring this optimised zero copy version as extension to API.  Then for now, I would explore if something needs to be done now, to allow easy extension of API (zero copy) down the line as non breaking change.

**Reply by NickUfer** [reply]
> Pinging @ncthbrt @NickUfer since they are interested in Rust bindings and this potential API change might make a difference there. I'm still a little hazy on best practices here for passing large arrays in and out of a library. What do other popular libraries do? In general I feel like a functional approach is nice where ownership isn't ever being passed around.

I think that whatever kind of zero copy optimizations you guys want to add should be fine with rust. The bindings are not done yet, but I already tested passing the vertProperties to rust and there was no problem. There is also no conversion going on, we can just use the data as is and construct a Vec out of it. Right now `manifold_meshgl_vert_properties` copies the data into a user provided buffer, but even if we just pass a pointer + length back to rust to achieve zero copy we can use the data as a Vec just fine without modification. This should hold true not only for vectors of primitive data types but also for structs. We simply mark them as `repr(C)` in rust, then they have the same order, size and alignment as in C/C++.

**Reply by starseeker** [reply]
Per https://github.com/elalish/manifold/issues/969#issuecomment-2391858849 sounds like we'll shelve the zero-copy question for now, and revisit (maybe with a parallel API) later if needed.

@elalish Should we set up an issue for the glm hiding so it can be added to the 3.0 milestone?

**Reply by elalish** [reply]
Sure, would be good to summarize the upshot of these long threads anyway.

**Comment by pca006132** [comment]
FYI: https://github.com/elalish/manifold/issues/976#issuecomment-2405681541
